### DISCUSSION TITLE
identify: %x format strings work, %[fx:resolution.x] outputs nothing

### DISCUSSION THREAD

**Discussion by acc987** [ImageMagick/ImageMagick#7622] [Help] [answered]
### ImageMagick version

7.1.1-21

### Operating system, version and so on

openSUSE Leap 15.5

### Description

I have a peculiar problem on the ImageMagick version available in the official openSUSE Leap 15.5 repositories:

For certain TIFF images, `identify -format "%x %y" IMAGE.TIFF` will always print the x and y resolutions as expected whereas `identify -format "%[fx:resolution.x] %[fx:resolution.y]" IMAGE.TIFF` will print only the whitespace between the %[] expressions.

There are some warnings when reading those images, but since %x and %y works I am not sure whether they are relevant:
```
identify: Unknown field with tag 18246 (0x4746) encountered. `TIFFReadDirectory' @ warning/tiff.c/TIFFWarnings/958.
identify: Unknown field with tag 18249 (0x4749) encountered. `TIFFReadDirectory' @ warning/tiff.c/TIFFWarnings/958.
identify: Incorrect count for "MakerNote"; tag ignored. `TIFFFetchNormalTag' @ warning/tiff.c/TIFFWarnings/958.
identify: Wrong data type 3 for "PixelXDimension"; tag ignored. `TIFFReadCustomDirectory' @ warning/tiff.c/TIFFWarnings/958.
identify: Wrong data type 3 for "PixelYDimension"; tag ignored. `TIFFReadCustomDirectory' @ warning/tiff.c/TIFFWarnings/958.
identify: Wrong data type 3 for "GainControl"; tag ignored. `TIFFReadCustomDirectory' @ warning/tiff.c/TIFFWarnings/958.
identify: ASCII value for tag "LensModel" contains null byte in value; value incorrectly truncated during reading due to implementation limitations. `TIFFFetchNormalTag' @ warning/tiff.c/TIFFWarnings/958.
identify: Unknown field with tag 18249 (0x4749) encountered. `TIFFReadDirectory' @ warning/tiff.c/TIFFWarnings/958.
identify: Unknown field with tag 18249 (0x4749) encountered. `TIFFReadDirectory' @ warning/tiff.c/TIFFWarnings/958.
```

Am I right to assume that this is an error with ImageMagick?

**Comment by snibgo** [comment]
> For certain TIFF images, ...

Please post an example.

**Comment by acc987** [comment]
It seems that proprietary software is responsible for creating the images I am having issues with and it took me a while to come up with a reproducer that I can share with you.

Anyway, the TIFF file in [SerraFina_AWFI_20200720_B16151413_8Bits_PL.zip](https://github.com/user-attachments/files/17032674/SerraFina_AWFI_20200720_B16151413_8Bits_PL.zip) reproduces the issue for me.

It's derived from the TIFF file at https://commons.wikimedia.org/wiki/File:SerraFina_AWFI_20200720_B16151413_8Bits.tiff (by [DPI/INPE](https://www.dpi.inpe.br/), licensed [CC BY-SA 4.0](https://creativecommons.org/licenses/by-sa/4.0/deed.en)) and thus also made available under the [Creative Commons Attribution-Share Alike 4.0 International](https://creativecommons.org/licenses/by-sa/4.0/deed.en) license. No changes were made to the image itself, it was simply re-exported as a TIFF file.

Interestingly, the image can be "fixed" by running `mogrify -strip` on it.

**Reply by fmw42** [reply]
Your tiff has two file in it. The first has density and units. The second has units but no density, so it likely will show the default density of 72.

I get on IM 7.1.1.31 on Mac OSX Ventura

```
magick identify -quiet -format "%x %y\n" SerraFina_AWFI_20200720_B16151413_8Bits_PL.tif
300 300
72 72
```

Which shows the results for the tiff with 2 files in it. The second shows the default of 72

`magick identify -quiet -format "%[fx:resolution.x] %[fx:resolution.y]\n" SerraFina_AWFI_20200720_B16151413_8Bits_PL.tif
 
 
`

which indeed shows nothing but the new lines.  **So there does seem to be a bug here as far as I can see.**

**Comment by snibgo** [comment]
The problem seems to be with "magick identify", but not with just "magick".

```
magick identify -quiet -format  "%[fx:resolution.x] %[fx:resolution.y]\n" SerraFina_AWFI_20200720_B1
6151413_8Bits_PL.tif


magick -quiet SerraFina_AWFI_20200720_B16151413_8Bits_PL.tif -format  "%[fx:resolution.x] %[fx:resolution.y]\n" info:
300 300
0 0
```
Tested with IM v7.1.1-20 and v7.1.1-38 on Windows 11.

**Reply by fmw42** [reply]
Interesting that you get 0 0 for the second image. The default should be 72 72.  So I guess the software for fx:resolutionx and fx:resolutiony  is different from %x, %y, when there is no value for the density (per magick identify -verbose)

**Comment by snibgo** [comment]
> So I guess the software for fx:resolution.x and fx:resolution.y is different from %x, %y

Yes.

For `resolution.x`, `property.` `GetMagickProperty()` returns `image->resolution.x`.

For `%x`, function `GetMagickPropertyLetter()` returns `image->resolution.x` if it is more than zero, otherwise it returns the default resolution.

**Comment by snibgo** [comment]
It seems that with `magick identify -format "%[fx:...]"`, the fx module is not called. So it never evaluates a `%[fx:...]` expression. For example:
```
magick identify -quiet -define fx:debug=true  -format  "Hello %[fx:23] world\n" SerraFina_AWFI_20200
720_B16151413_8Bits_PL.tif
Hello  world
Hello  world
```

**Reply by fmw42** [reply]
snibgo: Is that on the IM side or your fx side?

**Comment by snibgo** [comment, marked as answer]
fx.c has a bug.

When reading the supplied tiff, IM raises warning exceptions. The translation phase of fx sees the exceptions and stops processing and returns an error status. That is a bug: it should do so only when the exception is an error or worse.

Fixing the bug cures the OP problem.

The diff (from fx.c downloaded on 31 August 2024) to a version designated "AB" is:
```
--- "magickcore\\fx_20240831.c"	2024-07-21 12:11:21.754364200 +0100
+++ "magickcore\\fx_excep.c"	2024-09-18 03:28:27.241693900 +0100
@@ -2799,7 +2799,7 @@
     *needPopAll = MagickFalse;
   }
 
-  if (pfx->exception->severity != UndefinedException)
+  if (pfx->exception->severity >= ErrorException)
     return MagickFalse;
 
   return MagickTrue;
@@ -2851,7 +2851,7 @@
     }
   }
 
-  if (pfx->exception->severity != UndefinedException)
+  if (pfx->exception->severity >= ErrorException)
     return MagickFalse;
 
   return MagickTrue;
@@ -4034,7 +4034,7 @@
 
   if (NeedRelinq) cs = (ChannelStatistics *)RelinquishMagickMemory (cs);
 
-  if (pfx->exception->severity != UndefinedException) {
+  if (pfx->exception->severity >= ErrorException) {
     return MagickFalse;
   }
 
@@ -4382,7 +4382,7 @@
     }
   }
 
-  if (pfx->exception->severity != UndefinedException) {
+  if (pfx->exception->severity >= ErrorException) {
     status = MagickFalse;
   }

```

**Comment by dlemstra** [comment]
I just pushed a commit that includes the changes that you suggested Snibgo.

**Comment by snibgo** [comment]
Thanks.

**Comment by acc987** [comment]
Thanks to everybody involved for the analysis and the fix!

**Comment by snibgo** [comment]
The patch I showed above has not been fully implemented.

In fx.c, function FxImage(), line 4384, there are two lines:
```
  if (pfx->exception->severity != UndefinedException)
    status = MagickFalse;
```
But those lines should be:
```
  if (pfx->exception->severity >= ErrorException)
    status = MagickFalse;
```

**Reply by dlemstra** [reply]
I have applied a slightly different patch to fix this.
